[clang-scan-deps] Expand response files before the argument adjuster

aganea opened the pull request:
Previously, since response (.rsp) files weren't expanded at the very beginning of clang-scan-deps, we only parsed the command-line as provided in the Clang .cdb file. Unfortunately, when using Unreal Engine, arguments are always generated in a .rsp file (ie. `/path/to/clang-cl.exe @/path/to/filename_args.rsp`).

After this patch, `/Fo` can be parsed and added to the final command-line. Without this option, the make targets that are emitted are made up from the input file name alone. We have some cases where the same input in the project generates several output files, so we end up with duplicate make targets in the scan-deps emitted dependency file.

aganea committed:
[clang-scan-deps] Expand response files before the argument ajuster

Previously, since response (.rsp) files weren't expanded, we only parsed
the command-line as provided in the Clang CDB file. Unfortunately, when
using Unreal Engine, arguments are always generated in a .rsp file.

After this patch, /Fo can be parsed and added to the final command-line.
Without this option, the make targets that are emitted are made up from the
input file name alone. We have some cases where the same input in the project
generates several output files, so we end up with duplicate make targets
in the scan-deps emitted dependency file.

aganea commented:
This is missing a test, I will add one.

jansvoboda11 commented:
This change makes sense to me, but I forget what's the purpose of the whole output path deducing code. Even for `clang-cl` command lines we should go through the driver and get a `CompilerInvocation` that will have the output file set (if the action does produce an output file). Shouldn't that get propagated to the dependency consumer? Maybe the issue is that the `DependencyOptions` we pass have no target set, since the original command line doesn't have `-MD` (or something like that)?

aganea commented:
The reason of this `/Fo` output path fiddling code is this: https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/ToolChains/Clang.cpp#L1072-L1082
It's because that highlighted code doesn't handle the clang-cl case, it doesn't consider `/Fo`. Probably because most people don't use `-MT`/`-MF` with clang-cl (it has to be escaped on the command-line such as `/clang:-MT /clang:-MF...`. I can fix it there (in Clang.cpp) instead, and remove all the related code in `clang_scan_deps_main()`.

jansvoboda11 commented:
Hmm, that driver code is only executed if the dependency file was requested on the Clang command line, though. `clang-scan-deps` should be able to generate the dependency file even if the original command line doesn't have `-MD` and friends. That's why we call `deduceDepTarget()` in DependencyScanningWorker.cpp.

In general yes, I'd prefer to solve this wherever we have access to properly parsed command line (or `CompilerInvocation`) and removing the ad-hoc parsing in `clang-scan-deps`.

aganea committed:
Add test

aganea committed:
Handle /Fo in the Clang driver as well.

aganea commented:
@jansvoboda11 PTAL.

I've added handling of `/Fo` in the driver, and that solves my case without the .rsp expansion in `ClangScanDeps.cpp`. However if we pass `/E` to command-lines in the CDB, that change alone doesn't work anymore. This is because `Driver::GetNamedOutputPath()` returns "-" here: https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/Driver.cpp#L5871

@sylvain-audi worked around that by adding `-o file.o` from the adjuster callback, which happens to bypass that condition a bit earlier, here: https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/Driver.cpp#L5853

I'm not sure that's right, ideally we should pass `-MT` from the adjuster but somehow that doesn't work. This would probably need to be revisited later.

jansvoboda11 commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -1069,7 +1069,7 @@ void Clang::AddPreprocessingOptions(Compilation &C, const JobAction &JA,
       // If user provided -o, that is the dependency target, except
       // when we are only generating a dependency file.
-      Arg *OutputOpt = Args.getLastArg(options::OPT_o);
+      Arg *OutputOpt = Args.getLastArg(options::OPT_o, options::OPT__SLASH_Fo);
```
This change doesn't affect the test. Let's either add a test that exercises this, or drop this change.

aganea replied:
Added a test.

jansvoboda11 commented on clang/test/ClangScanDeps/response-file-clang-cl.c:
```diff
@@ -0,0 +1,36 @@
+
+// RUN: rm -rf %t
+// RUN: split-file %s %t
+// RUN: sed -e "s|DIR|%/t|g" %t/cdb.json.template > %t/cdb.json
```
You can now do this: `clang-scan-deps -o %t/deps.json -- %clang_cl <args>`. I guess `<args>` for `clang-cl` specifically cannot contain `%t` since that will begin with `/` on Unix system and cause command line parsing errors. But using relative paths like you currently do and changing directory before running the tools should work a will probably be much cleaner.

aganea replied:
As suggested. We're taking a different codepath when using the immediate mode (using `InplaceCompilationDatabase`) and that expands the response files in `OptTable::parseArgs()` so I wasn't really covering all the changes in the patch. I've added tests for both cases. The immediate mode allowed me to uncover a bug in `ParseArgs()`: `Saver` and `Alloc` were local to the function but we were using their contents to populate `CommandLine`. I fixed that too.

jansvoboda11 replied:
Ok, that makes sense, thanks!

jansvoboda11 commented on clang/test/ClangScanDeps/response-file-clang-cl.c:
```diff
@@ -0,0 +1,36 @@
+// RUN: rm -rf %t
+// RUN: split-file %s %t
+// RUN: sed -e "s|DIR|%/t|g" %t/cdb.json.template > %t/cdb.json
+// RUN: echo /Fo%t/tu.obj >> %t/args_nested.rsp
```
I tried playing around with response files and they don't play well with newlines. I think that instead of appending to the response files with `echo`, we should use `split-file` and create two separate files. That will also be more declarative and will make it obvious that one contains `/c` and the other `/E`.

aganea replied:
Can you elaborate please on "I tried playing around with response files and they don't play well with newlines"?
Things we working on my end (empty lines in .rsp files). Also, Unreal Engine generates .rsp files with each argument on a new line, so if you're seeing any issues there we shall fix them.

jansvoboda11 replied:
I recall running `%clang_cl` with whatever arguments and response files this test generated and seeing Clang's argument parser hit an assertion while working with the expanded arguments.

jansvoboda11 commented on clang/tools/clang-scan-deps/ClangScanDeps.cpp:
```diff
@@ -792,10 +792,15 @@ int clang_scan_deps_main(int argc, char **argv, const llvm::ToolContext &) {
 
+  // Expand response files in advance, so that we can "see" all the arguments
+  // when adjusting below.
+  auto ResponseExpander = expandResponseFiles(std::move(Compilations),
```
"Expander" sounds like a function object that you need to invoke to expand the compilations, while the result of `expandResponseFiles()` behaves like something that's already been expanded. I don't think the fact that this has been expanded is really relevant to the rest of the code here.

```suggestion
  Compilations = expandResponseFiles(std::move(Compilations),
```

aganea replied:
As suggested.

aganea committed:
Cover extracting -MT dependency from /Fo flag in clang-cl mode

aganea committed:
Address review comments

Also fix stack-allocated `CommandLine` entries when using a
InplaceCompilationDatabase

aganea commented:
@jansvoboda11 Do you see any further changes for this PR? Can I land it?

jansvoboda11 commented on clang/tools/clang-scan-deps/ClangScanDeps.cpp:
```diff
@@ -86,6 +86,8 @@ static bool DeprecatedDriverCommand;
 static ResourceDirRecipeKind ResourceDirRecipe;
 static bool Verbose;
 static bool PrintTiming;
+static llvm::BumpPtrAllocator Alloc;
+static llvm::StringSaver Saver{Alloc};
```
Why do these need to be global?

jansvoboda11 reviewed (approved):
LGTM, thanks!
